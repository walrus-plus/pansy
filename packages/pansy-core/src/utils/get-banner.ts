import { Banner } from '@pansy/types';
import { lodash } from '@walrus/shared-utils';
import stringifyAuthor from 'stringify-author';

/**
 * 获取Banner信息
 * @param banner banner配置
 * @param pkg package.json 中的数据
 */
export default (
  banner?: Banner,
  pkg?: { [k: string]: any }
): string => {
  if (!banner || lodash.isString(banner)) {
    return banner as string || '';
  }

  banner = { ...pkg, ...(banner === true ? {} : banner) };

  let author = '';

  if (lodash.isString(banner.author)) {
    author = banner.author;
  }

  if (lodash.isObject(banner.author)) {
    author = stringifyAuthor(banner.author)
  }

  const license = banner.license || '';

  return (
    '/*!\n' +
    ` * ${banner.name} v${banner.version}\n` +
    ` * (c) ${author || ''}\n` +
    (license && ` * Released under the ${license} License.\n`) +
    ' */'
  )
}
