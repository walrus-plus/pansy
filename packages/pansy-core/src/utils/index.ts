import { ModuleFormat as RollupFormat } from 'rollup'

export function getDefaultFileName(format: RollupFormat) {
  return format === 'cjs' ? `[name][min][ext]` : `[name].[format][min][ext]`
}
