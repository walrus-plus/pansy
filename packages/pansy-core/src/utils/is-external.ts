// @ts-ignore
import * as slash from 'slash';
import { lodash } from '@walrus/shared-utils';
import { Externals } from '@pansy/types';

export default function(externals: Externals, id: string, parentId?: string) {
  id = slash(id);

  if (!lodash.isArray(externals)) {
    externals = [externals] as Externals;
  }

  for (const external of externals) {
    if (
      lodash.isString(external) &&
      (id === external || id.includes(`/node_modules/${external}/`))
    ) {
      return true;
    }
    if (lodash.isRegExp(external)) {
      if (external.test(id)) {
        return true;
      }
    }
    if (typeof external === 'function') {
      if (external(id, parentId)) {
        return true;
      }
    }
  }

  return false;
}
