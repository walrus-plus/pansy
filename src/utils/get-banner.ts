import stringifyAuthor from 'stringify-author'

interface BannerInfo {
  /** Author name */
  name?: string
  /** package version */
  version?: string
  /** Author name or object */
  author?: any
  /** License name, like MIT */
  license?: string
}

export type Banner = string | BannerInfo | boolean

export default (banner?: Banner, pkg?: { [k: string]: any }): string => {
  if (!banner || typeof banner === 'string') {
    return banner || ''
  }

  banner = { ...pkg, ...(banner === true ? {} : banner) }

  const author =
    typeof banner.author === 'string'
      ? banner.author
      : typeof banner.author === 'object'
      ? stringifyAuthor(banner.author)
      : ''

  const license = banner.license || ''

  return (
    '/*!\n' +
    ` * ${banner.name} v${banner.version}\n` +
    ` * (c) ${author || ''}\n` +
    (license && ` * Released under the ${license} License.\n`) +
    ' */'
  )
}
