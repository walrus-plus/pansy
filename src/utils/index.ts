import { ModuleFormat } from 'rollup';
import { chalk as colors } from '@walrus/shared-utils';
import prettyBytes from 'pretty-bytes';
import boxen from 'boxen';
import textTable from 'text-table';
import stringWidth from 'string-width';
import logger from '../logger';

export interface Asset {
  absolute: string;
  source: string;
}

export type Assets = Map<string, Asset>;

export function getDefaultFileName(format: ModuleFormat) {
  return format === 'cjs' ? `[name][min][ext]` : `[name].[format][min][ext]`;
}

export async function printAssets(assets: Assets, title: string) {
  const gzipSize = await import('gzip-size').then((res) => res.default);
  const table = await Promise.all(
    [...assets.keys()].map(async (relative) => {
      const asset = assets.get(relative) as Asset;
      const size = asset.source.length;
      return [colors.green(relative), prettyBytes(size), prettyBytes(await gzipSize(asset.source))];
    })
  );
  table.unshift(['File', 'Size', 'Gzipped'].map((v) => colors.dim(v)));
  logger.success(title);
  logger.log(
    boxen(
      textTable(table, {
        stringLength: stringWidth
      })
    )
  );
}
