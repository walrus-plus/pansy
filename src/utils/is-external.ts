import slash from 'slash'
import { NormalizedConfig } from '../'

export default function(
  externals: NormalizedConfig['externals'],
  id: string,
  parentId?: string
) {
  id = slash(id)

  if (!Array.isArray(externals)) {
    externals = [externals] as NormalizedConfig['externals']
  }

  for (const external of externals) {
    if (
      typeof external === 'string' &&
      (id === external || id.includes(`/node_modules/${external}/`))
    ) {
      return true
    }
    if (external instanceof RegExp) {
      if (external.test(id)) {
        return true
      }
    }
    if (typeof external === 'function') {
      if (external(id, parentId)) {
        return true
      }
    }
  }

  return false
}
