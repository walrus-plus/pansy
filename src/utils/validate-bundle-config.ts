import signale from 'signale';
import { join } from 'path';
import isString from '@pansy/is-string';
import isArray from '@pansy/is-array';
import isObject from '@pansy/is-object';
import { existsSync } from 'fs';
import { chalk } from '@walrus/shared-utils';
import { isTypescriptFile } from './';
import { NormalizedConfig } from '../types';

export default function(config: NormalizedConfig, rootDir: string) {
  const { output, input } = config;

  if (output && output.format) {
    // 检验 cjs、commonjs是否重复设置
    if (output.format.indexOf('cjs') !== -1 && output.format.indexOf('commonjs') !== -1) {
      signale.warn(`${chalk.cyan('cjs')} and ${chalk.cyan('commonjs')} only need to set one`);
    }
    // 检验 es、esm是否重复设置
    if (output.format.indexOf('es') !== -1 && output.format.indexOf('esm') !== -1) {
      signale.warn(`${chalk.cyan('es')} and ${chalk.cyan('esm')} only need to set one`);
    }

    // umd格式必须制定output.name
    if (output.format.indexOf('umd') !== -1) {
      throw new Error(`${chalk.cyan('umd')} format must be set ${chalk.cyan('output.name')}`);
    }
  }

  const tsConfigPath = join(rootDir, 'tsconfig.json');
  const tsConfig = existsSync(tsConfigPath);

  let inputs: string[] = [];

  //  解析入口文件为数组
  if (input) {
    if (isString(input)) {
      inputs = [input];
    }
    if (isObject(input)) {
      Object.keys(input).forEach(key => {
        // @ts-ignore
        inputs.push(input[key]);
      })
    }
    if (isArray(config.input)) {
      inputs = config.input as string[];
    }
  }

  // ts项目没有配置tsconfig.json
  if (!tsConfig && inputs.some(isTypescriptFile as any)) {
    signale.info(
      `Project using ${chalk.cyan('typescript')} but tsconfig.json not exists. Use default config.`
    );
  }
}
