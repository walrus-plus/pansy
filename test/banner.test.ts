import { fixture, snapshot } from './utils';

/**
 * 开启Banner，自动读取package.json中的信息
 */
snapshot(
  {
    title: 'banner:true default',
    input: 'index.js',
    cwd: fixture('banner/default')
  },
  {
    banner: true
  }
);

/**
 * 直接指定Banner信息 BannerInfo
 */
snapshot(
  {
    title: 'banner:object',
    input: 'default.js',
    cwd: fixture()
  },
  {
    banner: {
      author: '阿康',
      license: 'MIT',
      name: 'pansy',
      version: '1.0.0'
    }
  }
);

/**
 * 直接指定Banner信息 string
 */
snapshot(
  {
    title: 'banner:string',
    input: 'default.js',
    cwd: fixture()
  },
  {
    banner: 'hello world'
  }
);
