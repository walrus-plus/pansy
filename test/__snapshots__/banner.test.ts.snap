// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`banner:object: banner:object dist/default.js 1`] = `
"/*!
 * pansy v1.0.0
 * (c) 阿康
 * Released under the MIT License.
 */
'use strict';

var _default = {};

module.exports = _default;
"
`;

exports[`banner:string: banner:string dist/default.js 1`] = `
"hello world
'use strict';

var _default = {};

module.exports = _default;
"
`;

exports[`banner:true default: banner:true default dist/index.js 1`] = `
"/*!
 * foo v0.0.1
 * (c) name <email@email.com>
 * Released under the MIT License.
 */
'use strict';

var index = 42;

module.exports = index;
"
`;
